plugin/evm/ftso_system.go

```diff
@@ -0,0 +1,268 @@
+	}
+
+	abiValidation, err := abi.JSON(strings.NewReader(jsonValidation))
+	if err != nil {
+		return nil, fmt.Errorf("could not get validation ABI: %w", err)
+	}
+
+	submitter := EVMContract{
+		address: addressSubmitter,
+		abi:     abiSubmitter,
+	}
+
+	validation := EVMContract{
+		address: addressValidation,
+		abi:     abiValidation,
+	}
+
+	abis := FTSOABIs{
+		Registry:  abiRegistry,
+		Manager:   abiManager,
+		Rewards:   abiRewards,
+		WNAT:      abiWNAT,
+		Whitelist: abiWhitelist,
+		Votepower: abiVotepower,
+	}
+
+	f := FTSOSystem{
+		blockchain: blockchain,
+		submitter:  submitter,
+		validation: validation,
+		abis:       abis,
+	}
+
+	return &f, nil
+}
+
+func (f *FTSOSystem) Contracts(hash common.Hash) (FTSOContracts, error) {
+
+	snap := BindEVM(f.blockchain).AtBlock(hash)
+
+	var registryAddress common.Address
+	err := snap.OnContract(f.submitter).Execute(RegistryAddress).Decode(&registryAddress)
+	if err != nil {
+		return FTSOContracts{}, fmt.Errorf("could not get registry address: %w", err)
+	}
+
+	registry := EVMContract{
+		address: registryAddress,
+		abi:     f.abis.Registry,
+	}
+
+	var managerAddress common.Address
+	err = snap.OnContract(f.submitter).Execute(ManagerAddress).Decode(&managerAddress)
+	if err != nil {
+		return FTSOContracts{}, fmt.Errorf("could not get manager address: %w", err)
+	}
+
+	manager := EVMContract{
+		address: managerAddress,
+		abi:     f.abis.Manager,
+	}
+
+	var rewardsAddress common.Address
+	err = snap.OnContract(manager).Execute(RewardsAddress).Decode(&rewardsAddress)
+	if err != nil {
+		return FTSOContracts{}, fmt.Errorf("could not get rewards address: %w", err)
+	}
+
+	rewards := EVMContract{
+		address: rewardsAddress,
+		abi:     f.abis.Rewards,
+	}
+
+	var whitelistAddress common.Address
+	err = snap.OnContract(f.submitter).Execute(WhitelistAddress).Decode(&whitelistAddress)
+	if err != nil {
+		return FTSOContracts{}, fmt.Errorf("could not get whitelist address: %w", err)
+	}
+
+	whitelist := EVMContract{
+		address: whitelistAddress,
+		abi:     f.abis.Whitelist,
+	}
+
+	var wnatAddress common.Address
+	err = snap.OnContract(rewards).Execute(WNATAddress).Decode(&wnatAddress)
+	if err != nil {
+		return FTSOContracts{}, fmt.Errorf("could not get WNAT address: %w", err)
+	}
+
+	wnat := EVMContract{
+		address: wnatAddress,
+		abi:     f.abis.WNAT,
+	}
+
+	var votepowerAddress common.Address
+	err = snap.OnContract(wnat).Execute(VotepowerAddress).Decode(&votepowerAddress)
+	if err != nil {
+		return FTSOContracts{}, fmt.Errorf("could not get votepower address: %w", err)
+	}
+
+	votepower := EVMContract{
+		address: votepowerAddress,
+		abi:     f.abis.Votepower,
+	}
+
+	contracts := FTSOContracts{
+		Registry:   registry,
+		Manager:    manager,
+		Rewards:    rewards,
+		Whitelist:  whitelist,
+		Votepower:  votepower,
+		Validation: f.validation,
+	}
+
+	return contracts, nil
+}
+
+func (f *FTSOSystem) Details(epoch uint64) (EpochDetails, error) {
```

Comment:
Why `epoch` isn't used anywhere in this function?

Reply:
It should be used as a parameter to the `Execute`.